File: lib/services/offline_semantic_encoder.dart
import 'dart:convert';
import 'dart:math' as math;

class OfflineSemanticEncoder {
  static const int dimension = 768;

  static const Set<String> _stopwords = {
    'a', 'an', 'and', 'are', 'as', 'at', 'be', 'by', 'for', 'from', 'have',
    'has', 'in', 'is', 'it', 'of', 'on', 'or', 'that', 'the', 'their', 'this',
    'to', 'was', 'were', 'will', 'with', 'you', 'your', 'i', 'we', 'they', 'them',
    'me', 'my', 'our', 'us', 'can', 'could', 'should', 'would', 'do', 'does', 'did',
    'not', 'no', 'yes', 'if', 'then', 'than', 'there', 'here', 'what', 'when',
    'where', 'how', 'why', 'which',
  };

  static const Map<String, List<String>> _synonyms = {
    'scheme': ['yojana', 'yojna', 'program', 'plan'],
    'apply': ['register', 'registration', 'application'],
    'register': ['apply', 'registration', 'enroll'],
    'free': ['zero', 'no-cost', 'without-payment'],
    'health': ['medical', 'hospital', 'clinic'],
    'hospital': ['clinic', 'medical', 'health'],
    'insurance': ['coverage', 'protection', 'benefit'],
    'treatment': ['care', 'medicine', 'therapy'],
    'medicine': ['drug', 'treatment', 'care'],
    'fever': ['bukhar', 'jvar'],
    'cough': ['khansi'],
    'pain': ['dard'],
    'help': ['support', 'assistance', 'madad'],
    'police': ['thana', 'law'],
    'woman': ['women', 'female'],
    'child': ['children', 'kid', 'baby'],
    'pregnant': ['pregnancy', 'mother'],
    'government': ['sarkar', 'govt'],
    'benefit': ['subsidy', 'support', 'aid'],
    'income': ['money', 'earnings'],
    'job': ['work', 'employment'],
    'water': ['drinking-water', 'safe-water'],
    'sanitation': ['toilet', 'hygiene'],
  };

  static List<int> encodeText(
    String text, {
    String? title,
    List<String>? keywords,
  }) {
    final vector = List<double>.filled(dimension, 0.0);

    final contentTokens = _tokenize(text);
    final titleTokens = _tokenize(title ?? '');
    final keywordTokens = (keywords ?? const <String>[])
        .map(_stem)
        .where((token) => token.isNotEmpty)
        .toList();

    _addTokenFeatures(vector, contentTokens, 1.0);
    _addTokenFeatures(vector, titleTokens, 1.35);
    _addTokenFeatures(vector, keywordTokens, 1.75);

    final combined = <String>[...contentTokens, ...titleTokens, ...keywordTokens];
    for (var index = 0; index < combined.length && index < 32; index++) {
      _addFeature(vector, 'pos:$index:${combined[index]}', 0.18);
    }

    final norm = _l2Norm(vector);
    if (norm == 0) {
      return List<int>.filled(dimension, 0);
    }

    final quantized = List<int>.filled(dimension, 0);
    for (var i = 0; i < vector.length; i++) {
      final scaled = ((vector[i] / norm) * 127.0).round();
      quantized[i] = scaled.clamp(-128, 127);
    }
    return quantized;
  }

  static List<String> _tokenize(String text) {
    final normalized = _normalizeText(text);
    if (normalized.isEmpty) return const <String>[];
    final tokens = normalized
        .split(' ')
        .where((token) => token.isNotEmpty && !_stopwords.contains(token))
        .toList();
    return tokens;
  }

  static String _normalizeText(String text) {
    final normalized = text
        .toLowerCase()
        .replaceAll(RegExp(r'[^a-z0-9]+'), ' ')
        .replaceAll(RegExp(r'\s+'), ' ')
        .trim();
    return normalized;
  }

  static String _stem(String token) {
    if (token.length <= 4) return token;
    for (final suffix in const ['ing', 'ed', 'es', 's']) {
      if (token.endsWith(suffix) && token.length - suffix.length >= 3) {
        return token.substring(0, token.length - suffix.length);
      }
    }
    return token;
  }

  static double _l2Norm(List<double> vector) {
    var total = 0.0;
    for (final value in vector) {
      total += value * value;
    }
    return math.sqrt(total);
  }

  static void _addTokenFeatures(
    List<double> vector,
    List<String> tokens,
    double baseWeight,
  ) {
    for (var index = 0; index < tokens.length; index++) {
      final token = tokens[index];
      if (token.isEmpty) continue;

      final base = _stem(token);
      _addFeature(vector, 'tok:$base', baseWeight);
      if (base != token) {
        _addFeature(vector, 'stem:$base', baseWeight * 0.85);
      }

      final synonymList = _synonyms[base];
      if (synonymList != null) {
        for (final synonym in synonymList) {
          _addFeature(vector, 'syn:$synonym', baseWeight * 0.7);
        }
      }

      for (final gram in _charTrigrams(base)) {
        _addFeature(vector, 'tri:$gram', baseWeight * 0.22);
      }

      if (index + 1 < tokens.length) {
        final bigram = '${base}_${_stem(tokens[index + 1])}';
        _addFeature(vector, 'bi:$bigram', baseWeight * 1.15);
      }
    }
  }

  static Iterable<String> _charTrigrams(String token) sync* {
    if (token.length < 3) return;
    for (var i = 0; i <= token.length - 3; i++) {
      yield token.substring(i, i + 3);
    }
  }

  static void _addFeature(List<double> vector, String feature, double weight) {
    final hash = _fnv1a32(utf8.encode(feature));
    final index = hash % dimension;
    final sign = (hash & 1) == 0 ? 1.0 : -1.0;
    vector[index] += weight * sign;
  }

  static int _fnv1a32(List<int> bytes) {
    var value = 0x811C9DC5;
    for (final byte in bytes) {
      value ^= byte;
      value = (value * 0x01000193) & 0xFFFFFFFF;
    }
    return value;
  }
}

File: assets/scripts/offline_semantic_encoder.py
"""Deterministic offline semantic encoder shared by the vector generator and test script.

This encoder is intentionally simple and fully offline:
- Normalize text
- Tokenize words
- Add stems, synonyms, title/keyword boosts, bigrams, and character trigrams
- Hash features into a fixed 768-dim vector using FNV-1a
- L2 normalize and quantize to int8

The same feature rules are mirrored in Dart so runtime query vectors live in the same space
as the precomputed vectors.json file.
"""

from __future__ import annotations

import math
import re
import unicodedata
from collections import defaultdict
from typing import Iterable, List, Sequence

DIMENSION = 768

TOKEN_RE = re.compile(r"[a-z0-9]+")

STOPWORDS = {
    "a", "an", "and", "are", "as", "at", "be", "by", "for", "from", "have",
    "has", "in", "is", "it", "of", "on", "or", "that", "the", "their", "this",
    "to", "was", "were", "will", "with", "you", "your", "i", "we", "they", "them",
    "me", "my", "our", "us", "can", "could", "should", "would", "do", "does", "did",
    "not", "no", "yes", "if", "then", "than", "there", "here", "what", "when",
    "where", "how", "why", "which",
}

SYNONYMS = {
    "scheme": ["yojana", "yojna", "program", "plan"],
    "apply": ["register", "registration", "application"],
    "register": ["apply", "registration", "enroll"],
    "free": ["zero", "no-cost", "without-payment"],
    "health": ["medical", "hospital", "clinic"],
    "hospital": ["clinic", "medical", "health"],
    "insurance": ["coverage", "protection", "benefit"],
    "treatment": ["care", "medicine", "therapy"],
    "medicine": ["drug", "treatment", "care"],
    "fever": ["bukhar", "jvar"],
    "cough": ["khansi"],
    "pain": ["dard"],
    "help": ["support", "assistance", "madad"],
    "police": ["thana", "law"],
    "woman": ["women", "female"],
    "child": ["children", "kid", "baby"],
    "pregnant": ["pregnancy", "mother"],
    "government": ["sarkar", "govt"],
    "benefit": ["subsidy", "support", "aid"],
    "income": ["money", "earnings"],
    "job": ["work", "employment"],
    "water": ["drinking-water", "safe-water"],
    "sanitation": ["toilet", "hygiene"],
}


def _normalize_text(text: str) -> str:
    text = unicodedata.normalize("NFKD", text)
    text = text.encode("ascii", "ignore").decode("ascii")
    text = text.lower()
    text = re.sub(r"[^a-z0-9]+", " ", text)
    return re.sub(r"\s+", " ", text).strip()


def _stem(token: str) -> str:
    if len(token) <= 4:
        return token
    for suffix in ("ing", "ed", "es", "s"):
        if token.endswith(suffix) and len(token) - len(suffix) >= 3:
            return token[: -len(suffix)]
    return token


def _tokenize(text: str) -> List[str]:
    normalized = _normalize_text(text)
    if not normalized:
        return []
    return [token for token in TOKEN_RE.findall(normalized) if token and token not in STOPWORDS]


def _fnv1a32(data: str) -> int:
    value = 0x811C9DC5
    for byte in data.encode("utf-8"):
        value ^= byte
        value = (value * 0x01000193) & 0xFFFFFFFF
    return value


def _add_feature(vector: List[float], feature: str, weight: float) -> None:
    hash_value = _fnv1a32(feature)
    index = hash_value % DIMENSION
    sign = 1.0 if (hash_value & 1) == 0 else -1.0
    vector[index] += weight * sign


def _char_ngrams(token: str, n: int = 3) -> Sequence[str]:
    if len(token) < n:
        return ()
    return tuple(token[i : i + n] for i in range(len(token) - n + 1))


def build_features(
    text: str,
    title: str | None = None,
    keywords: Iterable[str] | None = None,
) -> List[float]:
    vector = [0.0] * DIMENSION

    def add_token_features(tokens: Sequence[str], base_weight: float) -> None:
        for index, token in enumerate(tokens):
            if not token:
                continue
            base = _stem(token)
            _add_feature(vector, f"tok:{base}", base_weight)
            if base != token:
                _add_feature(vector, f"stem:{base}", base_weight * 0.85)

            for synonym in SYNONYMS.get(base, ()):
                _add_feature(vector, f"syn:{synonym}", base_weight * 0.7)

            for gram in _char_ngrams(base):
                _add_feature(vector, f"tri:{gram}", base_weight * 0.22)

            if index + 1 < len(tokens):
                bigram = f"{base}_{_stem(tokens[index + 1])}"
                _add_feature(vector, f"bi:{bigram}", base_weight * 1.15)

    content_tokens = _tokenize(text)
    title_tokens = _tokenize(title or "")
    keyword_tokens = [_stem(token) for token in (keywords or ()) if token]

    add_token_features(content_tokens, 1.0)
    add_token_features(title_tokens, 1.35)
    add_token_features(keyword_tokens, 1.75)

    combined = content_tokens + title_tokens + keyword_tokens
    for index, token in enumerate(combined[:32]):
        _add_feature(vector, f"pos:{index}:{token}", 0.18)

    # Normalize and quantize later.
    return vector


def encode_text(text: str, title: str | None = None, keywords: Iterable[str] | None = None) -> List[int]:
    vector = build_features(text, title=title, keywords=keywords)
    norm = math.sqrt(sum(value * value for value in vector))
    if norm == 0:
        return [0] * DIMENSION
    quantized = []
    for value in vector:
        scaled = int(round((value / norm) * 127.0))
        if scaled < -128:
            scaled = -128
        elif scaled > 127:
            scaled = 127
        quantized.append(scaled)
    return quantized


def encode_text_float(text: str, title: str | None = None, keywords: Iterable[str] | None = None) -> List[float]:
    vector = build_features(text, title=title, keywords=keywords)
    norm = math.sqrt(sum(value * value for value in vector))
    if norm == 0:
        return [0.0] * DIMENSION
    return [value / norm for value in vector]
